src/Compilers/CSharp/Portable/Symbols/Synthesized/InjectedNonNullTypesAttributeSymbol.cs

```diff
@@ -0,0 +1,188 @@
+            {
+                // member is missing
+                var memberDescriptor = WellKnownMembers.GetDescriptor(WellKnownMember.System_AttributeUsageAttribute__ctor);
+                var diagnosticInfo = new CSDiagnosticInfo(ErrorCode.ERR_MissingPredefinedMember, memberDescriptor.DeclaringTypeMetadataName, memberDescriptor.Name);
+                Diagnostics.Add(diagnosticInfo, Location.None);
+
+                return ImmutableArray<CSharpAttributeData>.Empty;
+            }
+
+            Binder.ReportUseSiteDiagnostics(ctor, Diagnostics, Location.None);
+
+            NamedTypeSymbol attributeTargets = DeclaringCompilation.GetWellKnownType(WellKnownType.System_AttributeTargets);
+            Binder.ReportUseSiteDiagnostics(attributeTargets, Diagnostics, Location.None);
+
+            var usage = new SynthesizedAttributeData(ctor,
+                arguments: ImmutableArray.Create(new TypedConstant(attributeTargets, TypedConstantKind.Enum, attributeUsage)),
+                namedArguments: ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+
+            return ImmutableArray.Create<CSharpAttributeData>(usage);
+        }
+
+        internal override AttributeUsageInfo GetAttributeUsageInfo()
+            => new AttributeUsageInfo(validTargets: attributeUsage, allowMultiple: false, inherited: false);
+
+        private const AttributeTargets attributeUsage =
+            AttributeTargets.Class |
+            AttributeTargets.Constructor |
+            AttributeTargets.Delegate |
+            AttributeTargets.Enum |
+            AttributeTargets.Event |
+            AttributeTargets.Field |
+            AttributeTargets.Interface |
+            AttributeTargets.Method |
+            AttributeTargets.Module |
+            AttributeTargets.Property |
+            AttributeTargets.Struct;
+
+        private sealed class NonNullTypesAttributeConstructorSymbol : SynthesizedInstanceConstructor
+        {
+            private readonly ImmutableArray<ParameterSymbol> _parameters;
+
+            internal NonNullTypesAttributeConstructorSymbol(
+                NamedTypeSymbol containingType,
+                Func<MethodSymbol, ImmutableArray<ParameterSymbol>> getParameters) :
+                base(containingType)
+            {
+                Debug.Assert(containingType is InjectedAttributeSymbol);
+                _parameters = getParameters(this);
+            }
+
+            public override ImmutableArray<ParameterSymbol> Parameters
+                => _parameters;
+
+            internal override bool SynthesizesLoweredBoundBody
+                => true;
+
+            /// <summary>
+            /// Note: this method captures diagnostics into the containing type (an injected attribute symbol) instead,
+            /// as we don't yet know if the containing type will be emitted.
+            /// </summary>
+            internal override void GenerateMethodBody(TypeCompilationState compilationState, DiagnosticBag diagnostics)
```

Comment:
>GenerateMethodBody [](start = 35, length = 18)

This looks identical to the corresponding method for `EmbeddedAttribute`. Consider extracting a helper method, perhaps in `SynthesizedInstanceConstructor`.

Reply:
The NonNullTypes attribute will end up different (it needs to save the parameter value into a field), so I kept a separate implementation.

This also differs from `SynthesizedEmbeddedAttributeConstructorSymbol` in its handling of diagnostics. One uses the diagnostic bag parameter, but the injected symbols use the diagnostic bag of the containing type.

Reply:
As we discussed, the helper method could have parameters for diagnostics and a callback for any additional statements.

---
In reply to: [214428014](https://github.com/dotnet/roslyn/pull/29180#discussion_r214428014) [](ancestors = 214428014)